File: XRViewer/ARKController/SceneKit/PlaneNode.m
#import "PlaneNode.h"

@interface PlaneNode ()
@end


@implementation PlaneNode

- (instancetype)initWithAnchor:(ARPlaneAnchor *)anchor
{
    self = [super init];
    
    if (self)
    {
        [self setup];
        
        [self update:anchor];
    }
    
    return self;
}

- (void)update:(ARPlaneAnchor *)anchor
{
    SCNPlane *plane = (SCNPlane *)[self geometry];
    [plane setWidth:anchor.extent.x];
    [plane setHeight:anchor.extent.z];
    
    SCNMaterial *material = [plane firstMaterial];
    [[material diffuse] setContentsTransform:SCNMatrix4MakeScale(anchor.extent.x, anchor.extent.z, 1)];
    [[material diffuse] setWrapS:SCNWrapModeRepeat];
    [[material diffuse] setWrapT:SCNWrapModeRepeat];
    
    SCNMatrix4 translateTransform = SCNMatrix4MakeTranslation(anchor.center.x, 0, anchor.center.z);
    SCNMatrix4 rotationTransform = SCNMatrix4MakeRotation(-M_PI / 2.0, 1.0, 0.0, 0.0);    

    [self setTransform:SCNMatrix4Mult(rotationTransform, translateTransform)];
}

- (void)setup
{
    SCNPlane *plane = [SCNPlane new];
    
    SCNMaterial *material = [SCNMaterial new];
    UIImage *img = [UIImage imageNamed:@"Models.scnassets/plane_grid1.png"];
    [[material diffuse] setContents:img];
    [[material diffuse] setIntensity:0.5];
    [plane setMaterials:@[material]];
    
    [self setGeometry:plane];
    [self setTransform:SCNMatrix4MakeRotation(-M_PI / 2.0, 1, 0, 0)];
    
    [self setOpacity:0];
}

@end

File: XRViewer/ARKController/SceneKit/PlaneNode.swift
import ARKit
import SceneKit

class PlaneNode: SCNNode {
    init(anchor: ARPlaneAnchor?) {
        super.init()

        setup()
        update(anchor)
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    func update(_ anchor: ARPlaneAnchor?) {
        guard let anchor = anchor else { return }
        let plane = geometry as? SCNPlane
        plane?.width = CGFloat(anchor.extent.x)
        plane?.height = CGFloat(anchor.extent.z)

        let material: SCNMaterial? = plane?.firstMaterial
        material?.diffuse.contentsTransform = SCNMatrix4MakeScale(anchor.extent.x, anchor.extent.z, 1)
        material?.diffuse.wrapS = .repeat
        material?.diffuse.wrapT = .repeat

        let translateTransform: SCNMatrix4 = SCNMatrix4MakeTranslation(anchor.center.x, 0, anchor.center.z)
        let rotationTransform: SCNMatrix4 = SCNMatrix4MakeRotation(-.pi / 2.0, 1.0, 0.0, 0.0)
        transform = SCNMatrix4Mult(rotationTransform, translateTransform)
    }

    func setup() {
        let plane = SCNPlane()

        let material = SCNMaterial()
        let img = UIImage(named: "Models.scnassets/plane_grid1.png")
        material.diffuse.contents = img
        material.diffuse.intensity = 0.5
        plane.materials = [material]

        geometry = plane
        transform = SCNMatrix4MakeRotation(-.pi / 2.0, 1, 0, 0)
        opacity = 0
    }
}
